File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/StatisticsSample/src/com/hp/workpath/sample/statisticsample/dialog/CommitDialogFragment.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.statisticsample.dialog;

import android.app.Activity;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.NumberPicker;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;

import com.hp.workpath.sample.statisticsample.Logger;
import com.hp.workpath.sample.statisticsample.R;
import com.hp.workpath.sample.statisticsample.task.CommitTask;
import com.hp.workpath.sample.statisticsample.task.LastJobSequenceTask;

public class CommitDialogFragment extends DialogFragment implements LastJobSequenceTask.LastJobSequenceTaskCompletionListener,CommitTask.CommitTaskListener {

    private static int totalTaskCount;
    private TextView mLastJobSequenceTextView;
    private TextView mTotalTextView;
    private NumberPicker mNumberPicker;

    public static CommitDialogFragment newInstance(int totalTaskCount) {
        CommitDialogFragment.totalTaskCount = totalTaskCount;
        return new CommitDialogFragment();
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        LayoutInflater inflater = getActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.dialog_commit, null);
        findViewElements(view);

        AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(getActivity())
                .setTitle(R.string.commit)
                .setView(view)
                .setPositiveButton(android.R.string.ok, mOKListener)
                .setNegativeButton(android.R.string.cancel, mCancelListener)
                .setCancelable(false);
        return dialogBuilder.create();
    }

    private void findViewElements(View view) {
        mLastJobSequenceTextView = view.findViewById(R.id.lastJobSequenceTextView);
        mTotalTextView = view.findViewById(R.id.totalTextView);
        mNumberPicker = view.findViewById(R.id.numberPicker);
        new LastJobSequenceTask(getContext(), this).taskExecute();

    }

    private final DialogInterface.OnClickListener mOKListener = new DialogInterface.OnClickListener() {
        @Override
        public void onClick(DialogInterface dialog, int which) {
            try {
                int commit = mNumberPicker.getValue();
                new CommitTask(getContext(), commit,CommitDialogFragment.this).taskExecute();
            } catch (Throwable t) {
                Logger.showResult(getActivity(), "StatisticsService.commit " + t.getMessage());
            }
            dialog.dismiss();
        }
    };

    private final DialogInterface.OnClickListener mCancelListener = new DialogInterface.OnClickListener() {
        @Override
        public void onClick(DialogInterface dialog, int which) {
            dialog.cancel();
        }
    };

    @Override
    public void onLastJobSequenceCompleted(int lastJobSequence) {
        try {
            mLastJobSequenceTextView.setText(getString(R.string.committed_job_sequence, lastJobSequence));
            mTotalTextView.setText(getString(R.string.total, totalTaskCount));
            mNumberPicker.setDescendantFocusability(NumberPicker.FOCUS_BLOCK_DESCENDANTS);


            if (totalTaskCount > 0) {
                mNumberPicker.setMinValue(lastJobSequence + 1);
                mNumberPicker.setMaxValue(lastJobSequence + totalTaskCount);
            } else {
                mNumberPicker.setEnabled(false);
            }
        } catch (Throwable t) {
            Logger.showResult(getActivity(), "LastJobSequenceTask, TotalCountTask " + t.getMessage());
        }
    }

    @Override
    public void onCommitTask(Boolean isCommitted) {
        Logger.showResult(getActivity(), "commit result: " + isCommitted);
        getTargetFragment().onActivityResult(getTargetRequestCode(), Activity.RESULT_OK, null);
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/StatisticsSample/src/com/hp/workpath/sample/statisticsample/dialog/CommitDialogFragment.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.statisticsample.dialog

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.View
import android.widget.NumberPicker
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.core.os.bundleOf
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.lifecycleScope
import com.hp.workpath.sample.statisticsample.Logger
import com.hp.workpath.sample.statisticsample.R
import com.hp.workpath.sample.statisticsample.StatisticActivity.Companion.REQUEST_KEY
import com.hp.workpath.sample.statisticsample.task.CommitTask
import com.hp.workpath.sample.statisticsample.task.LastJobSequenceTask
import com.hp.workpath.sample.statisticsample.task.TotalCountTask
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking

class CommitDialogFragment : DialogFragment() {

    private lateinit var mLastJobSequenceTextView: TextView
    private lateinit var mTotalTextView: TextView
    private lateinit var mNumberPicker: NumberPicker

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val inflater = requireActivity().layoutInflater
        val view = inflater.inflate(R.layout.dialog_commit, null)
        findViewElements(view)
        val dialogBuilder = AlertDialog.Builder(requireActivity())
            .setTitle(R.string.commit)
            .setView(view)
            .setPositiveButton(android.R.string.ok, mOKListener)
            .setNegativeButton(android.R.string.cancel, mCancelListener)
            .setCancelable(false)
        return dialogBuilder.create()
    }

    private fun findViewElements(view: View) = runBlocking {
        mLastJobSequenceTextView = view.findViewById(R.id.lastJobSequenceTextView)
        mTotalTextView = view.findViewById(R.id.totalTextView)
        mNumberPicker = view.findViewById(R.id.numberPicker)

        try {
            val lastJobSequenceDeferred = lifecycleScope.async(Dispatchers.Default) {
                context?.let { LastJobSequenceTask(it).execute() }
            }
            val lastJobSequence = lastJobSequenceDeferred.await() ?: 0
            mLastJobSequenceTextView.text = getString(R.string.committed_job_sequence, lastJobSequence)

            val totalCountDeferred = lifecycleScope.async(Dispatchers.Default) {
                context?.let { TotalCountTask(it).execute() }
            }
            val totalCount = totalCountDeferred.await()
            mTotalTextView.text = getString(R.string.total, totalCount)
            mNumberPicker.descendantFocusability = NumberPicker.FOCUS_BLOCK_DESCENDANTS

            if (totalCount != null && totalCount > 0) {
                mNumberPicker.minValue = lastJobSequence + 1
                mNumberPicker.maxValue = lastJobSequence + totalCount
            } else {
                mNumberPicker.isEnabled = false
            }
        } catch (t: Throwable) {
            Logger.showResult(activity, "LastJobSequenceTask, TotalCountTask ${t.message}")
        }
    }

    private val mOKListener = DialogInterface.OnClickListener { dialog, _ ->
        try {
            runBlocking {
                val commit = mNumberPicker.value
                val isCommitted = lifecycleScope.async(Dispatchers.Default) {
                    context?.let { CommitTask(it, commit).execute() }
                }
                Logger.showResult(activity, "commit result: ${isCommitted.await()}")
                parentFragmentManager.setFragmentResult(
                    REQUEST_KEY, // Same request key StoreJobFragment used to register its listener
                    bundleOf() // The data to be passed to StoreJobFragment
                )
            }
        } catch (t: Throwable) {
            Logger.showResult(activity, "StatisticsService.commit ${t.message}")
        }
        dialog.dismiss()
    }
    private val mCancelListener =
        DialogInterface.OnClickListener { dialog, _ -> dialog.cancel() }

    companion object {
        fun newInstance(): CommitDialogFragment {
            return CommitDialogFragment()
        }
    }
}